### ISSUE TITLE
NullPointerException (io.socket.client.Socket.onevent(Unknown Source))

### ISSUE THREAD

**Issue by syrogath** [socketio/socket.io-client-java#505]
Hey there!

I'm working on an android/ios app game working with this library and a node.js backend also working with socket.io. On the ios app, everything works well, but my analytics tool shows me that in android there are a lot of users facing a fatal exception. Here it is:

```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
       at io.socket.client.Socket.onevent(Unknown Source)
       at io.socket.client.Socket.connect(Unknown Source)
       at io.socket.client.Socket.open(Unknown Source)
       at io.socket.client.Socket$2$2.call(Unknown Source)
       at io.socket.emitter.Emitter.on(Unknown Source)
       at io.socket.client.Manager.emitAll(Unknown Source)
       at io.socket.client.Manager.reconnection(Unknown Source)
       at io.socket.client.Manager$7.call(Unknown Source)
       at io.socket.emitter.Emitter.on(Unknown Source)
       at io.socket.parser.Parser$Decoder.add(Unknown Source)
       at io.socket.client.Manager.emitAll(Unknown Source)
       at io.socket.client.Manager.reconnection(Unknown Source)
       at io.socket.client.Manager$2.call(Unknown Source)
       at io.socket.emitter.Emitter.on(Unknown Source)
       at io.socket.engineio.client.Socket.open(Unknown Source)
       at io.socket.engineio.client.Socket.open(Unknown Source)
       at io.socket.engineio.client.Socket$5.call(Unknown Source)
       at io.socket.emitter.Emitter.on(Unknown Source)
       at io.socket.engineio.client.Transport.onError(Unknown Source)
       at io.socket.engineio.client.transports.Polling.pause(Unknown Source)
       at io.socket.engineio.client.transports.Polling$2.call(Unknown Source)
       at io.socket.engineio.parser.Parser.encodePacket(Unknown Source)
       at io.socket.engineio.client.transports.Polling.pause(Unknown Source)
       at io.socket.engineio.client.transports.Polling.pause(Unknown Source)
       at io.socket.engineio.client.transports.PollingXHR$5$1.run(Unknown Source)
       at io.socket.thread.EventThread$2.run(Unknown Source)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
       at java.lang.Thread.run(Thread.java:762)
```
I'm using version 1.7.4 on node.js backend, version 0.9.0 from this library. I tried many things getting rid of this exception, but nothing seems to work like using this library in a singleton or in a service.

Is there anyone facing this problem too?

If you need more code examples of the project or want to have a look on the app (it's in the stores), let me know.

**Comment by JarnoVgr** [comment]
Can you provide your actual code? (relevant pieces)

**Comment by syrogath** [comment]
This is my Service I'm using with this library. I also tried out using socket.io in a singleton and in application class. But nothing fixes the exception. The URL is a https one.

```
public class SocketService extends Service {

    public class LocalBinder extends Binder {
        public SocketService getService() { return SocketService.this; }
    }

    private final IBinder binder = new LocalBinder();
    private Socket socket;

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        try {
            IO.Options options = new IO.Options();
            options.forceNew = true;

            socket = IO.socket(ApiConstants.Routes.URL, options);
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }

        socket.on(Socket.EVENT_CONNECT, (dataArray) -> {
            Log.e("SocketManager", "connected");

            if (Prefs.getAuthToken() != null && !Prefs.getAuthToken().isEmpty()) {
                JSONArray array = new JSONArray();
                array.put(Prefs.getAuthToken());

                if (Prefs.getGroupId() != null && !Prefs.getGroupId().isEmpty()) {
                    array.put(Prefs.getGroupId());
                }

                socket.emit("updateUserSocket", array);
            } else {
                socket.disconnect();
                socket.off();
            }
        });

        socket.on(Socket.EVENT_DISCONNECT, (dataArray) -> {
            Log.e("SocketManager", "disconnected");
            sendBroadcast(Socket.EVENT_DISCONNECT);
        });

        socket.on(Constants.SocketEvents.USER_UPDATED, (dataArray) -> {
            Log.e("SocketManager", "user updated");
            sendBroadcast(Constants.SocketEvents.USER_UPDATED);
        });

        socket.on(Constants.SocketEvents.USER_SELF_REMOVED, (dataArray) -> {
            Log.e("SocketManager", "user self removed");
            sendBroadcast(Constants.SocketEvents.USER_SELF_REMOVED);
        });

        socket.on(Constants.SocketEvents.START_GAME, (dataArray) -> {
            Log.e("SocketManager", "start game");
            sendBroadcast(Constants.SocketEvents.START_GAME);
        });

        socket.on(Constants.SocketEvents.START_VOTING, (dataArray) -> {
            Log.e("SocketManager", "start voting");
            sendBroadcast(Constants.SocketEvents.START_VOTING);
        });

        socket.on(Constants.SocketEvents.GROUP_DELETED, (dataArray) -> {
            Log.e("SocketManager", "group deleted");
            sendBroadcast(Constants.SocketEvents.GROUP_DELETED);
        });

        socket.on(Constants.SocketEvents.UPDATE_GROUP, (dataArray) -> {
            Log.e("SocketManager", "update group");
            sendBroadcast(Constants.SocketEvents.UPDATE_GROUP);
        });

        socket.on(Constants.SocketEvents.GAME_P1, (dataArray) -> {
            Log.e("SocketManager", "game p1");
            sendBroadcast(Constants.SocketEvents.GAME_P1);
        });

        socket.on(Constants.SocketEvents.GAME_P2, (dataArray) -> {
            Log.e("SocketManager", "game p2");
            sendBroadcast(Constants.SocketEvents.GAME_P2);
        });

        socket.on(Constants.SocketEvents.GAME_P3, (dataArray) -> {
            Log.e("SocketManager", "game p3");
            sendBroadcast(Constants.SocketEvents.GAME_P3);
        });

        socket.on(Constants.SocketEvents.GAME_P4, (dataArray) -> {
            Log.e("SocketManager", "game p4");
            sendBroadcast(Constants.SocketEvents.GAME_P4);
        });

        socket.connect();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return Service.START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        socket.disconnect();
        socket.off();
    }

    private void sendBroadcast(String event) {
        Intent intent = new Intent(event);
        LocalBroadcastManager.getInstance(IncognyApp.getContext()).sendBroadcast(intent);
    }
}
```

**Comment by vooda--** [comment]
We have same issue. Does anybody have solution yet?

**Comment by iagormoraes** [comment]
> We have same issue. Does anybody have solution yet?

Bit late, but check the reply [explaining what is happening with versions 1.0.1 and lower](https://github.com/socketio/socket.io-client-java/issues/642#issuecomment-820843371)

**Comment by darrachequesne** [comment]
For future readers:

This should be fixed by https://github.com/socketio/socket.io-client-java/commit/e8ffe9d1383736f6a21090ab959a2f4fa5a41284, which was included in version 2.0.1.

This commit was also [backported](https://github.com/socketio/socket.io-client-java/commit/8664499b6f31154f49783531f778dac5387b766b) to the 1.x branch, and included in version `1.0.2`.

Please reopen if needed.

**Comment by rmirabelle** [comment]
Not sure if this is the same bug, but I'm running 2.1.0:

```
java.lang.NullPointerException · Attempt to read from field 'int io.socket.parser.Packet.id' on a null object reference

Socket.java:232 io.socket.client.Socket$6$1.runTimer.java:562 
java.util.TimerThread.mainLoopTimer.java:512 
java.util.TimerThread.run
```
